### PR TITLE
Add ability to switch off the animation of showons

### PR DESCRIPTION
**Pull request by regularlabs**
### Summary of Changes
To conditionally hide fields or content based on the values of other fields, you can add a `showon=""` attribute to fields in xml files.

However, this will always use a slideUp/slideDown animation.
There are cases where you do not want to animate the showing/hiding of your content or fields.

This patch adds the ability to switch off the animation by adding a class`no-animation` or `no-animate` to the field.
You can also add the class to the surrounding showon container, which can be helpful when creating your own via php.

### Testing Instructions
For instance, you can test with the Joomla core configuration:
```
		<field
			name="page_heading"
			type="text"
			label="COM_MENUS_ITEM_FIELD_PAGE_HEADING_LABEL"
			description="COM_MENUS_ITEM_FIELD_PAGE_HEADING_DESC"
			default=""
			showon="show_page_heading:1"
		/>
```
And change that to:
```
		<field
			name="page_heading"
			type="text"
			label="COM_MENUS_ITEM_FIELD_PAGE_HEADING_LABEL"
			description="COM_MENUS_ITEM_FIELD_PAGE_HEADING_DESC"
			default=""
			showon="show_page_heading:1"
			class="no-animation"
		/>
```
Now go to: `administrator/index.php?option=com_config&view=component&component=com_menus` and toggle the `Show Page Heading` setting.
Before the patch it should have the slide animation. After, it should not animate.

### Documentation Changes Required
https://docs.joomla.org/Form_field
Optionally add a note somewhere that you can switch off the animation effect via the classname.

### CODE REVIEW

File: media/jui/js/cms-uncompressed.js [hunk 1]
```diff
@@ -137,6 +141,10 @@ Joomla = window.Joomla || {};
 				}
 			}
 
+			if(typeof animate === "undefined") {
```
Comment (Quy) [inline media/jui/js/cms-uncompressed.js]:
```suggestion
			if (typeof animate === "undefined") {
```

File: media/jui/js/cms-uncompressed.js [hunk 2]
```diff
@@ -61,6 +61,10 @@ Joomla = window.Joomla || {};
 				fieldName = condition.field;
 				$fields   = $('[name="' + fieldName + '"], [name="' + fieldName + '[]"]');
 
+				if(typeof animate === "undefined" && typeof condition.animate !== "undefined") {
```
Comment (Quy) [inline media/jui/js/cms-uncompressed.js]:
```suggestion
				if (typeof animate === "undefined" && typeof condition.animate !== "undefined") {
```

**Comment by brianteeman** [comment]
It does what it says but I dont see the use case??

**Comment by regularlabs** [comment]
The slide animation is vertical.
There are cases when this messes up the layout during animation.
Especially when using custom fields that have different layouts than the standard label=>field.

**Comment by brianteeman** [comment]
Can you share an example?

**Comment by regularlabs** [comment]
With animation:
with-animation

Without Animation:
no-animation

**Comment by brianteeman** [comment]
Thanks

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 8f31b56fa9ed6b2ffafe68a62d7637b1911626a5<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22870">issues.joomla.org/tracker/joomla-cms/22870</a>.</sub>

**Comment by regularlabs** [comment]
Also there are cases when you have a LOT of fields hiding/showing based on a single value.
And then animation can cause the page to react very slowly (especially on older browsers).

Also, visually, animation it is not always what you want.
For instance, when you want to toggle 2 fields (same label & options, just different field name) without the user seeing it is actually a different field.
For instance, an active and a readonly (disabled) version of the field.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 8f31b56fa9ed6b2ffafe68a62d7637b1911626a5<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22870">issues.joomla.org/tracker/joomla-cms/22870</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22870">issues.joomla.org/tracker/joomla-cms/22870</a>.</sub>

**Comment by bembelimen** [comment]
Any chance to offer more possibilities as value for this parameter? "none", "vertical", "horizontal", ...

**Comment by dgrammatiko** [comment]
FWIW in Joomla 4 there ARE NO animations, so introducing this feature in staging means that the project should keep the same API (with the redundant animation part) also in J4. @wilsonge keep a note on this please...

**Comment by regularlabs** [comment]
@bembelimen 
There is no ready-to-use horizontal slide effects in the jQuery version included in Joomla 3.
So I think it is beyond the scope of this PR to implement that.

**Comment by regularlabs** [comment]
Any reason why this isn't merged yet?

**Comment by dgrammatiko** [comment]
@regularlabs I guess because of 
https://github.com/joomla/joomla-cms/pull/22870#issuecomment-434618117

**Comment by regularlabs** [comment]
No, because of:
https://github.com/joomla/joomla-cms/pull/22912#issuecomment-438895546

**Comment by dgrammatiko** [comment]
Honestly it will be ultra stupid to introduce an API change in 3.9 and remove it in 4.0...

**Comment by regularlabs** [comment]
It is solving an issue we have in Joomla 3.
If Joomla 4 solves that in another fashion (by having no animations at all), then fine.
Joomla 4 is changing behaviour anyway.

**Comment by dgrammatiko** [comment]
@regularlabs I think you miss my point, you can implement this without adding technical debt for J4 by using a class, eg `no-animation`. This will have minimal impact for J4, in contrary the introduction of another attribute if that needs to be removed in J4 brings a B/C break...

**Comment by regularlabs** [comment]
I totally agree with you!

I have changed the code, which means that only the cms.js file has now changed.
No php changes. Not API changes in the xml elements.

I also updated the description of the PR.

So I guess the RTC tag needs to get removed and we need to do testing again.

If it is better to trash this PR and create a new fresh one, let me know.

**Comment by zero-24** [comment]
tanking RTC off as requested.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22870">issues.joomla.org/tracker/joomla-cms/22870</a>.</sub>

**Comment by regularlabs** [comment]
@Quy Could you test again? See if it is still working as it should?

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on a85f10c45fc84b338c69dc0ad86b928e0ad93f4b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22870">issues.joomla.org/tracker/joomla-cms/22870</a>.</sub>

**Comment by dgrammatiko** [comment]
I have tested this item :white_check_mark: successfully on a85f10c45fc84b338c69dc0ad86b928e0ad93f4b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22870">issues.joomla.org/tracker/joomla-cms/22870</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22870">issues.joomla.org/tracker/joomla-cms/22870</a>.</sub>

**Comment by regularlabs** [comment]
👍

**Comment by C-Lodder** [comment]
One thing I would have suggested in this is not to use `toggle()` as it was deprecated in 1.8 and remove in 1.9. The only reason it works is because you ship jquery-migrate, however there are some optimisation plugins that allow you to unset it from the document and thus `toggle()` will be undefined.

Should have used `.css(...)` instead.

**Comment by regularlabs** [comment]
The existing js code there was already using toggle(). This PR hasn't changed that.

**Pull request merged**